### ISSUE TITLE
Remove implicit magic from module __init__.py

### ISSUE THREAD

**Issue by fonfon** [freedombox/Plinth#447]
We have a lot of implicit and magic functionality happening around the modules `__init__.py` files. Lots of it is neither obvious for newcomers nor for me if I didn't look at the code for a while.

If #446 is merged, the remaining implicit variables are: `version, depends, service, managed_services, title, description, init, setup, diagnose, is_essential`.

It's hard to find out how and where such things are used, and it's error-prone because there may be some that you don't know about and set accidentially. The python zen says:

> Explicit is better than implicit.

 I'd like to keep this 'signature' of a module to the absolute minimum, and to make it as explicit as possible. Perhaps we can make the module declaration (setup etc.) explicit, perhaps we can provide something like an 'api' object that combines all those 'signature' variables/functions.

**Comment by SunilMohanAdapa** [comment]
I don't believe this falls into 'Explicit is better than implicit'.

Currently, all of the above functions are acting like an API for a application. #446 devolves this API.  Earlier there is a consistent way in which one can query whether an application is enabled (or if it is running, or a way to enable/disable). This could have been used in a dashboard.  Now there isn't.  Removing these existing and consistent APIs does not make the code more explicit.

Imagine a class called Application which every application derives from and implements the above functions and properties.  That would very similar to how Django does it's various classes.  Now a module (instead of a class) can be thought of as a singleton class (with some disadvantages). This is what we have now.

I agree that the current code is less readable.  The solution I would propose is to have every application derive from an App base class and provide properties and override its methods:
```
class App():
  version = 1
  depends = []
  is_essential = False
  title = None
  description = None
  packages = []
  services = []

  def init()
  def is_setup()
  def is_enabled()
  def enable(should_enable)
  def setup() 
  def diagnose()
```

**Comment by SunilMohanAdapa** [comment]
@fonfon , does the plan to convert Apps into classes solve the readability problem? Are you interested in picking this up?

**Comment by fonfon** [comment]
Yes this sounds like a good idea, and I would like to pick it up.

It can take a while until I find some time, so if somebody wants to implement this earlier please let me know, or just assign this issue to yourself too.

**Comment by SunilMohanAdapa** [comment]
@fonfon Good to know you are working this issue again.  Recently, there is an introduction of framework to add shortcuts into the front page.  This mostly shows up as `add_shortcut()` in a given module called from `init()`, `enable()` and `disable()`. Perhaps `add_shortcut()` could be made part of the base class and automatically called at appropriate times giving the app implementer the ability to override methods when alteration is needed.

Apart from that an uninstall method is also submitted as a pull request.

Please consider post any changes you think should happen to the API here before putting in effort to change code.

**Comment by fonfon** [comment]
I made a rough architecture draft to demonstrate and discuss the changes as a pull request:
https://github.com/fonfon/Plinth/pull/1

The proposed changes lie in between the `Apps` class you proposed, and djangos `AppConfig` and `App` pattern. I made more detailed comments in the pull request and in the code.

The pull request shows what the (draft) classes and API could look like, and how they are used by the deluge module. It's far from complete or elaborate, just to show the direction.

**Comment by SunilMohanAdapa** [comment]
I wasn't aware of the AppConfig class in Django. Your comparison of the two classes is quite apt and the central idea of the patch seems to be in the right direction.

Deriving from AppConfig it not a bad idea:
- Actually moving the application class into modules/module/apps.py has some advantages such as avoiding unnecessary import problems.
- If for some reason a developer has to create a Django AppConfig, it gets messy since there will be very similar properties such as title, id, etc. of the app that will get duplicated.
- Even if we don't derive from AppConfig, we could still reuse its API, and convention of keeping it in apps.py for simpler understanding on part of a Django developer.
- I have also suggested using the term 'App' instead of 'Module' at the end of this response.

> class PlinthModuleManager

The name 'Plinth' could be removed. In this case, it seems to be used to distinguish between the base ModuleManager and application's ModuleManager.  However, it is preferable to use namespaces to distinguish different classes and use import syntax and resolve the ambiguity in the same module.

So, plinth.module.ModuleManager is the base class and plinth.modules.deluge.ModuleManager is the derived class. Code in deluge could look like:
```python
import plinth.module

class ModuleManager(plinth.module.ModuleManager):
```
or
```python
from plinth.module import ModuleManager as PlinthModuleManager

class ModuleManager(PlinthModuleManager):
``` 

> class PlinthModuleManager

It is typically a convention in software development to use 'XManager' to refer to a supervisor object that handles a bunch of instances of X.

In our case, it will at least mean that there is actually a Module that is managed by this ModuleManager.  This will require additional understanding on behalf of the application developer without actual need for two level handling.

This class could called simply `class Module`. The 'Modules' class can be called 'ModuleManager' since it is not a simple collection of modules and is actually doing the duty of instantiating the modules, etc. All of the functionality of the module_loader.py could be part ModuleManager.

> version = 0

- Properties must be heavily documented as these are now API.

- Version should default to 1 if we intend to make the life of application developer somewhat easy with the defaults. Otherwise, it should default to None and we should raise an exception such as ImpropertyConfigured in `__init__.py` of Module class.

- `__init__.py` should check and raise exceptions for values that are compulsory and not provided properly.

> service = None

This should probably become an instance variable rather than class variable.

> managed_services = ['deluge-web']
> managed_packages = ['deluged', 'deluge-web']

The names are `services` and `packages` as defined in the API.

> class PlinthModuleManager

This class should implement most of the API methods.
- These methods will define the API that every module should conform to.
- If a module does not wish to handle that particular aspect of implementation, it will leave that to default implementation.  For example, a module which does not wish to implement diagnostics will simply ignore it and may rely on Module.diagnose() to implement some basic diagnostics.
- All the callers of the application, such as the diagnostics module, need to be assured of the presence of a particular method (such as diagnose()).  We can avoid the current strategy of checking if a method is present and then calling it.

> class ModuleSetup(object):
> """
> Setup functionality for a module (same as the Helper class below).

Renaming this module does not seem very necessary.

> It provides all setup-related functionality and information of a module,
> and is accessible via a ModuleManager instance: <module_manager>.setup

Nice. 

> setup_modules and list_dependencies could/should be part of the Modules class

If everything related to <module>.diagnose is related to diagnostics and is managed by diagnostics module, then everything related to <module>.packages should be managed by the setup module as it is doing currently. So, these two methods should stay part of the setup functionality instead of becoming module manager functionality.

> plinth/modules/__init__.py

plinth.modules is a currently a [implicit namespace package](https://www.python.org/dev/peps/pep-0420/). Creating the `__init__.py` will convert it into a regular package.  This means that it will no longer be possible to have sub-modules of this packaged placed in separate locations on the disk.

I suggest having a plinth/module.py instead.

> class PlinthModuleManager(object):

The term 'module' is an artifact of the pre-Django days of Plinth.  We should transition to the term App or Application instead:

- This is consistent with Django's terminology.  Explaining to developers that there is a Plinth module which like a Django app will get confusing.  Saying that Plinth application is a Django application (with additional API) will make them much more comfortable.
- The term 'module' has an implication that the main application is what is important and that the module enhances its functionality. On the other hand, 'application' gives the idea of an independently useful functionality and that Plinth is merely as enabler and a helper. I believe this is true for what Plinth does.
- We wish to promote FreedomBox app development as if promoting Mobile application development. We wish people not associated with the core project developing applications for FreedomBox. Since in the current industry, Mobile app development is popular, the term 'app' is relatable and conveys what developer should be doing. If we use term FreedomBox 'app' in marketing and then talk about Plinth modules and Django apps, things will be confusing. Instead FreedomBox app == Plinth App =~ Django App is much better for understanding.

We should try to move away from the term 'module' and embrace the term 'app' instead. In future we can rename plinth/modules to plinth/apps.
